### ISSUE TITLE
Does exoplayer detect if HDMI is UP on android TV?

### ISSUE THREAD

**Issue by bene25** [google/ExoPlayer#4817]
We have a problem with exoplayer playback at start after sleep mode. Our Android TV app are set as default on STB. When you launch STB it doesnt show you launcher, it shows you our App immediately with playing HLS video stream. For playing video we are using Exo Player version 2.6.1 with Exo Media wrapper version 4.1.0. Time by time we investigate black screen when STB wakes up after sleep - power On by DPAD. Playlist what we get in this cases is working and switching channel to another works well. This issue is produced time by time commonly at the same users.

Here what we get from log when screen is black (URL_TEST is our playlist request):

```
09-06 19:21:40.018 I/PowerManagerService(  369): Waking up from sleep...
09-06 19:21:40.018 I/WindowManager(  369): No lock screen!
09-06 19:21:40.038 I/AudioHardwareALSA(  104): AudioHardwareALSA::setParameters() screen_state=on
09-06 19:21:40.038 I/qtaguid (  369): Failed write_ctrl(s 1 10056) res=-1 errno=1
09-06 19:21:40.038 W/NetworkManagementSocketTagger(  369): setKernelCountSet(10056, 1) failed with errno -1
09-06 19:21:40.078 D/LightsService(  369): Excessive delay setting light: 54ms
09-06 19:21:41.018 W/SurfaceFlinger(  101): Timed out waiting for hw vsync; faking it
09-06 19:21:41.158 W/OkHttp  (  682): previous digest authentication with same nonce failed, returning null
09-06 19:21:41.158 D/OkHttp  (  682): Cached authentication expired. Sending a new request.
09-06 19:21:42.018 W/SurfaceFlinger(  101): Timed out waiting for hw vsync; faking it
09-06 19:21:42.098 D/DAKARMainActivity(  682): URL_TEST getUri: http://XXX.XXX.XX.XXX:XXXXX/XLIVE/0431233...8355920521/26/3/stream.m3u8
09-06 19:21:43.098 W/SurfaceFlinger(  101): Timed out waiting for hw vsync; faking it
09-06 19:21:43.428 W/DisplayD(   97): iface id HDMI screen is 0
09-06 19:21:43.428 D/Displaymanager(   97): [setHDMIEnable] display 0
09-06 19:21:43.428 D/Displaymanager(   97): [operateIfaceMode] property 0 iface 4 type 1 mode 1280x720p-60
09-06 19:21:43.428 D/Displaymanager(   97): [operateIfaceEnable] property 0 iface 4 operate 1 value 1
09-06 19:21:43.438 W/Displaymanager(   97): interface 4 not support led
09-06 19:21:43.648 I/ActivityManager(  369): Config changes=1400 {1.0 ?mcc?mnc en_LU ldltr sw720dp w1280dp h672dp 160dpi lrg long land -touch -keyb/v/h dpad/v s.8}
09-06 19:21:43.658 D/RemoteControlManager(  369): BroadcastReceiver:android.intent.action.CONFIGURATION_CHANGED
```
What we got from log when everything is OK and video is playing after start on the same device (URL_TEST is our playlist request):

```
09-05 19:21:58.507 I/PowerManagerService(  369): Waking up from sleep...
09-05 19:21:58.517 I/WindowManager(  369): No lock screen!
09-05 19:21:58.527 D/SurfaceFlinger(  102): Screen acquired, type=0 flinger=0x40691318
09-05 19:21:58.537 I/AudioHardwareALSA(  105): AudioHardwareALSA::setParameters() screen_state=on
09-05 19:21:58.537 I/qtaguid (  369): Failed write_ctrl(s 1 10056) res=-1 errno=1
09-05 19:21:58.537 W/NetworkManagementSocketTagger(  369): setKernelCountSet(10056, 1) failed with errno -1
09-05 19:21:58.577 W/DisplayD(   98): iface id HDMI screen is 0
09-05 19:21:58.577 D/Displaymanager(   98): [setHDMIEnable] display 0
09-05 19:21:58.577 D/Displaymanager(   98): [operateIfaceMode] property 0 iface 4 type 1 mode 1280x720p-60
09-05 19:21:58.577 D/Displaymanager(   98): [operateIfaceEnable] property 0 iface 4 operate 1 value 1
09-05 19:21:58.577 W/Displaymanager(   98): interface 4 not support led
09-05 19:21:58.587 D/LightsService(  369): Excessive delay setting light: 59ms
09-05 19:21:58.797 D/RemoteControlManager(  369): BroadcastReceiver:android.intent.action.CONFIGURATION_CHANGED
09-05 19:21:58.797 D/Environment(  369): screen raw size:1280,720
09-05 19:21:58.797 I/ActivityManager(  369): Config changes=1400 {1.0 ?mcc?mnc en_LU ldltr sw720dp w1280dp h672dp 160dpi lrg long land -touch -keyb/v/h dpad/v s.9}
09-05 19:21:58.947 W/OkHttp  (  681): previous digest authentication with same nonce failed, returning null
09-05 19:21:58.947 D/OkHttp  (  681): Cached authentication expired. Sending a new request.
09-05 19:21:59.077 I/Displaymanager(   98): updatesinkaudioinfo: 0x408de140
09-05 19:21:59.077 I/Displaymanager(   98): BUF: AC3,LPCM,
09-05 19:21:59.077 I/Displaymanager(   98): 
09-05 19:21:59.077 W/DisplayD(   98): Send msg Iface added HDMI
09-05 19:21:59.137 D/DAKARMainActivity(  681): URL_TEST getUri: http://XXX.XXX.XX.XXX:13888/XLIVE/4086274310...086993824321/26/3/stream.m3u8
09-05 19:21:59.417 I/OMXClient(  681): Using client-side OMX mux.
09-05 19:21:59.417 I/        (  681): new RKOMXPlugin
09-05 19:21:59.427 D/ACodec  (  681): MEDIA_MIMETYPE_AUDIO_AAC sampleRate 48000 wifidisplay_flag 0 info 6c5bbe78
09-05 19:21:59.427 E/OMXNodeInstance(  681): OMX_GetExtensionIndex failed   
```
We noticed from log when we have black screen we have this Displaymanager messages after video request:

```
 09-06 19:21:42.098 D/DAKARMainActivity(  682): URL_TEST getUri: http://XXX.XXX.XX.XXX:XXXXX/XLIVE/0431233...8355920521/26/3/stream.m3u8
09-06 19:21:43.098 W/SurfaceFlinger(  101): Timed out waiting for hw vsync; faking it
09-06 19:21:43.428 W/DisplayD(   97): iface id HDMI screen is 0
09-06 19:21:43.428 D/Displaymanager(   97): [setHDMIEnable] display 0
09-06 19:21:43.428 D/Displaymanager(   97): [operateIfaceMode] property 0 iface 4 type 1 mode 1280x720p-60
09-06 19:21:43.428 D/Displaymanager(   97): [operateIfaceEnable] property 0 iface 4 operate 1 value 1
09-06 19:21:43.438 W/Displaymanager(   97): interface 4 not support led
```
When everything is OK it goes vice versa - see logs above.

Our assumption that black screen could be caused by exoplayer, which doesnt check if HDMI service is  UP or not at the moment when we request video stream.

We have tried to detect HDMI status by all of this methods (when HDMI is connected/up then we start video playing) - https://stackoverflow.com/questions/8644665/how-to-check-the-hdmi-device-connection-status-in-android/, but nothing helps. Also we tried to make DisplayManager.DisplayListener, it also doesnt work. 

Android version 17.

Any ideas what could cause this problem?
The same issue on stackoverflow - https://stackoverflow.com/questions/52312407/detect-if-hdmi-service-is-up-on-android-tv

**Comment by ojw28** [comment]
We don't detect the HDMI status anywhere in the player, but it's somewhat unclear that we should need to. For video output ExoPlayer just outputs video to any valid `Surface`. It doesn't really need to care about HDMI. For audio output we're aware that playback can currently fail on an HDMI status change, but you should be seeing an obvious error propagated to `EventListener.onPlayerError` in that case. Do you see this? If that's the problem, I think @andrewlewis is planning to address that.

Please can you try registering an `EventLogger` with the player instance to add some more useful logging? And then capturing a full bug report as requested in the issue template just after the error is encountered?

**Comment by bene25** [comment]
We use this listener from [ExoMedia](https://github.com/brianwernick/ExoMedia) to catch errors, and there are no error:

```
package com.devbrackets.android.exomedia.listener;

/**
 * Interface definition of a callback to be invoked when there
 * has been an error during an asynchronous operation.
 */
public interface OnErrorListener {
    /**
     * Called to indicate an error has occurred. In the case that the error was
     * caused by the Native {@link android.media.MediaPlayer} <code>e</code> should
     * be of the type {@link com.devbrackets.android.exomedia.core.exception.NativeMediaPlaybackException}
     * while all other exceptions will be provided by the {@link com.google.android.exoplayer2.ExoPlayer}
     * implementation directly.
     *
     * @param e The {@link Exception} that caused or describes the error
     * @return True if the method handled the error, false if it didn't.
     */
    boolean onError(Exception e);
}
```

**Comment by andrewlewis** [comment]
Do you get an `AudioManager.ACTION_HDMI_AUDIO_PLUG` event when the screen turns on in the case were output is blank? If not, do you receive this in the case where the output isn't blank? (If you don't already have a listener for this broadcast, ExoPlayer's `AudioCapabilitiesReceiver` shows how to register one.)

**Comment by bene25** [comment]
> We don't detect the HDMI status anywhere in the player, but it's somewhat unclear that we should need to. For video output ExoPlayer just outputs video to any valid `Surface`. It doesn't really need to care about HDMI. For audio output we're aware that playback can currently fail on an HDMI status change, but you should be seeing an obvious error propagated to `EventListener.onPlayerError` in that case. Do you see this? If that's the problem, I think @andrewlewis is planning to address that.

I have logged 

```
private void handlePlayerError(final ExoPlaybackException e) {
      eventListenerHandler.post(new Runnable() {
        @Override
        public void run() {
          for (Player.EventListener listener : eventListeners) {
            Log.d("TAG", "TestLog: "+ e);
            listener.onPlayerError(e);
          }
        }
      });
      changePlaybackState(Player.STATE_ENDED);
    }

```
and 

```
 case ExoPlayerImplInternal.MSG_ERROR: {
        ExoPlaybackException exception = (ExoPlaybackException) msg.obj;
        for (Player.EventListener listener : listeners) {
          Log.d(TAG, "TestLog: "+ exception);
          listener.onPlayerError(exception);
        }
        break;
      }

```
But nothing at log, while black screen appeared.


> Please can you try registering an `EventLogger` with the player instance to add some more useful logging? And then capturing a full bug report as requested in the issue template just after the error is encountered?

We have exo player version  2.6.1 which was without EventLogger class. In our version we have made some locally modifications. And it takes some time to update version to the latest. I have tried to insert EventLogger manually but to much classes have to be added. Could you tell what methods you want to log and I will do it. Thanks!

**Comment by ojw28** [comment]
Regarding logging, you could copy the `EventLogger` that was in the demo app in 2.6.1, [here](https://github.com/google/ExoPlayer/blob/r2.6.1/demos/main/src/main/java/com/google/android/exoplayer2/demo/EventLogger.java) and use that. You need to register it as multiple listener types, as shown [here](https://github.com/google/ExoPlayer/blob/r2.6.1/demos/main/src/main/java/com/google/android/exoplayer2/demo/PlayerActivity.java#L301).

**Comment by bene25** [comment]
> Do you get an `AudioManager.ACTION_HDMI_AUDIO_PLUG` event when the screen turns on in the case were output is blank? If not, do you receive this in the case where the output isn't blank? (If you don't already have a listener for this broadcast, ExoPlayer's `AudioCapabilitiesReceiver` shows how to register one.)

Did it. 
But ACTION_HDMI_AUDIO_PLUG requires API level 21, we have 17.

**Comment by andrewlewis** [comment]
I thought Android TV required API 21 (Lollipop) or later. Could you paste the output of `adb shell getprop ro.build.fingerprint`?

**Comment by bene25** [comment]
> I thought Android TV required API 21 (Lollipop) or later. Could you paste the output of `adb shell getprop ro.build.fingerprint`?

rk31sdk/rk31sdk/rk31sdk:4.2.2/JDQ39/eng.root.20140122.105724:eng/test-keys

**Comment by bene25** [comment]
Here is log with EventLogger.
First TEST2 appearing is black screen, second TEST2 appearing is working channel (switching to another channel) 

[log.txt](https://github.com/google/ExoPlayer/files/2400316/log.txt)

**Comment by bene25** [comment]
Hi guys. Any suggestions what could cause this problem?

**Comment by andrewlewis** [comment]
Could you file an issue with the OEM providing the device? It sounds like a device-specific issue and I don't think there's much we can do: it seems there's no error being surfaced to ExoPlayer, and I'm not aware of any APIs you can use to wait until HDMI output is initialized on API 17.

**Comment by bene25** [comment]
While investigation, we found that after black screen video resumes itself after some time.
[video_resuming.txt](https://github.com/google/ExoPlayer/files/2405973/video_resuming.txt)

**Comment by bene25** [comment]
We made more tests, and added small 2 seconds delay, but nothing changes. Black screen still appear.
[black_screen_2.txt](https://github.com/google/ExoPlayer/files/2415399/black_screen_2.txt)

**Comment by bene25** [comment]
Thanks for helping.
Problem resolved with ExoPlayer update to 2.8.4 and ExoMedia to 4.2.3
